src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+        if (mpi_rank == chunk_entry->new_owner) {
             if (chunk_entry->num_writers > 1) {
                 MPI_Message message;
                 MPI_Status  status;
                 size_t      j;
 
-                chunk_entry->async_info.num_receive_requests = (int)chunk_entry->num_writers - 1;
-                if (NULL == (chunk_entry->async_info.receive_requests_array = (MPI_Request *)H5MM_malloc(
+                /* Allocate all necessary buffers for a non-blocking receive operation */
+                chunk_entry->async_info.num_receive_requests = chunk_entry->num_writers - 1;
+                if (NULL == (chunk_entry->async_info.receive_requests_array = H5MM_malloc(
                                  (size_t)chunk_entry->async_info.num_receive_requests * sizeof(MPI_Request))))
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate async requests array")
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate MPI requests array")
 
                 if (NULL ==
-                    (chunk_entry->async_info.receive_buffer_array = (unsigned char **)H5MM_malloc(
+                    (chunk_entry->async_info.receive_buffer_array = H5MM_malloc(
                          (size_t)chunk_entry->async_info.num_receive_requests * sizeof(unsigned char *))))
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate async receive buffers")
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate receive buffers")
 
-                for (j = 0; j < chunk_entry->num_writers - 1; j++) {
+                for (j = 0; j < (size_t)chunk_entry->num_writers - 1; j++) {
                     int count = 0;
 
-                    /* Probe for a particular message from any process, removing that message
-                     * from the receive queue in the process and allocating that much memory
-                     * for the asynchronous receive
+                    /* Probe for a particular message for this chunk from any rank, removing
+                     * that message from the receive queue for the rank and allocating that
+                     * much memory for the non-blocking receive
                      */
-                    if (MPI_SUCCESS != (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index,
-                                                              io_info->comm, &message, &status)))
+                    if (MPI_SUCCESS !=
+                        (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index_info.chunk_idx,
+                                               io_info->comm, &message, &status)))
                         HMPI_GOTO_ERROR(FAIL, "MPI_Mprobe failed", mpi_code)
 
                     if (MPI_SUCCESS != (mpi_code = MPI_Get_count(&status, MPI_BYTE, &count)))
                         HMPI_GOTO_ERROR(FAIL, "MPI_Get_count failed", mpi_code)
 
                     HDassert(count >= 0);
```

Comment:
So, I agree with this point and did change it to an error check instead of an assert in my current set of changes. That said, I think the only danger here is if the MPI implementation is buggy/broken, in which case we probably have bigger problems. But yes, an assert probably wasn't best to use here, so I switched it (final changes to come in another PR rather than making this one larger).